deduction_rules.py

```diff
@@ -167,4 +167,162 @@ def line_square(sudoku):
                     r,c=local_to_global(sec,i,j)
                     if c!=col:
                         made_deduction|=sudoku.ban(r,c,val+1,"col_square",reason)
+    return made_deduction
+
+def naked_trios(sudoku):
+    """RULE: if v,w,x can be written only to three cells in this row/col/sec, then remove it from the other cells of the row/col/sec."""
+
+    def search_for_naked_trios(elems):
+        """Returns a list of 3-tuples, where every tuple contains the indeces of the identical inputs. The cases:
+        (123) (123) (123) - {3/3/3} (in terms of candidates per cell)
+        (123) (123) (12) - {3/3/2} (or some combination thereof)
+        (123) (12) (23) - {3/2/2/}
+        (12) (23) (13) - {2/2/2}
+        """
+        l = []
+        for comb in combinations(range(9),3):
+            cell0 = set(elems[comb[0]])
+            cell1 = set(elems[comb[1]])
+            cell2 = set(elems[comb[2]])
+            for cell0, cell1, cell2 in permutations((cell0, cell1, cell2)):
+                if len(cell0) == 3 and cell0 == cell1 and cell1 == cell2:
+                    l.append(tuple(sorted((comb[0],comb[1],comb[2]))))
+                    break
+                elif len(cell0) == 3 and cell0 == cell1 and len(cell2) == 2 and cell2.issubset(cell1):
+                    l.append(tuple(sorted((comb[0],comb[1],comb[2]))))
+                    break
+                elif len(cell0) == 3 and len(cell1) == 2 and len(cell2) == 2 and \
+                    cell1.issubset(cell0) and cell2.issubset(cell0) and len(cell1.intersection(cell2)) == 1:
+                    l.append(tuple(sorted((comb[0],comb[1],comb[2]))))
+                    break
+                elif len(cell0) == 2 and len(cell1) == 2 and len(cell1) == 2 and \
+                    len(set.intersection(cell0,cell1)) == 2 and len(set.intersection(cell1,cell2)) == 2 and \
+                    len(set.intersection(cell0,cell2)) == 2 and len(set.intersection(cell0,cell1,cell2)) == 1:
+                    l.append(tuple(sorted((comb[0],comb[1],comb[2]))))
+                    break
+        return l
+    
+    def search_and_ban_in_subset(cells_to_check):
+        """Searches all naked-trios in a subset of cells and bans these numbers from the other elements of subset."""
+        made_deduction = False
+        allowed_numbers = _allowed_numbers(sudoku,cells_to_check)
+        naked_trios = search_for_naked_trios(allowed_numbers)
+        for trio in naked_trios:
+            print("naked_trios :)")
+            cell0 = cells_to_check[trio[0]]
+            cell1 = cells_to_check[trio[1]]
+            cell2 = cells_to_check[trio[2]]
+            deleted_numbers = set(sudoku.allowed[cell0[0]][cell0[1]].allowed() + sudoku.allowed[cell1[0]][cell1[1]].allowed() + sudoku.allowed[cell2[0]][cell2[1]].allowed())
+            cells_used = sudoku.allowed[cell0[0]][cell0[1]].notNones()+sudoku.allowed[cell1[0]][cell1[1]].notNones()+sudoku.allowed[cell2[0]][cell2[1]].notNones()
```

Comment:
This once again conflicts with the docstring of  `naked_pairs`, but confroms to the original definition.